### DISCUSSION TITLE
Morph Marker injection inside HTML element (v3)

### DISCUSSION THREAD

**Discussion by thumbtech** [livewire/livewire#7389] [7. Bugs (archived)]
### Livewire version

v3.2.1

### Which PHP version are you using?

PHP 8.2

### Steps To Reproduce

Morph markers are injected into the following HTML in blade:

```
      <button type="button" wire:click="incrDate(-1)" class="
        @if ($date > $today)
        text-blue-900 sm:hover:bg-blue-100
        @else
        text-blue-300
        @endif
        rounded-l-md relative inline-flex items-center bg-blue-50 px-2 py-2 focus:z-10
        "
        @if ($date <= $today)
        title="Past Disabled"
        disabled
        @else
        title="Previous Day's Orders"
        @endif
        > 
```

morphed into (not exactly ... copied from inspector which messes it up a bit ... but suffice to say busted):

```
<button type="button" wire:click="incrDate(-1)" class="
                text-blue-300
         <!--[if ENDBLOCK]><![endif]-->
        rounded-l-md relative inline-flex items-center bg-blue-50 px-2 py-2 focus:z-10
        " <!--[if="" block]=""><!--[endif]---->        title="Past Disabled"
        disabled
                &gt;
```

FWIW I've tried to reverse engineer the directive / pattern stuff in src/Features/SupportMorphAwareIfStatement/SupportMorphAwareIfStatement.php in order to try a regex search in my blade templates using Visual Studio and better assess the risk of this feature on my existing code base (mature & working fine in v2). Definitely not a regex whiz and haven't gotten good results. I'm definitely gun shy. I've also had it inject markers into a JS script on the blade template.

### Please confirm (incomplete submissions will not be addressed)

- [X] I have provided easy and step-by-step instructions to reproduce the bug.
- [X] I have provided code samples as text and NOT images.
- [X] I understand my bug report will be removed if I haven't met the criteria above.

**Comment by thumbtech** [comment]
Follow up on regex search in VS Code on my project (searching in ./resources/views)

Exact PHP-generated opening patterns pasted into VS Code finds 0 results:
```
            (@foreach(?![a-zA-Z])|@forelse(?![a-zA-Z])|@unless(?![a-zA-Z])|@switch(?![a-zA-Z])|@error(?![a-zA-Z])|@isset(?![a-zA-Z])|@empty(?![a-zA-Z])[^\S\r
]*\(|@guest(?![a-zA-Z])|@while(?![a-zA-Z])|@auth(?![a-zA-Z])|@for(?![a-zA-Z])|@if(?![a-zA-Z]))  # Blade directives: (@if|@foreach|...)
            (?!                     # Not followed by:
                [^<]*               # ...
                (?<![?=-])          # ... (Make sure we don't confuse ?>, ->, and =>, with HTML opening tag closings)
                >                   # A ">" character that isn't preceded by a "<" character (meaning it's outside of a tag)
            )
```

removing 1st newline (only this!) returns 573 results:
```
            (@foreach(?![a-zA-Z])|@forelse(?![a-zA-Z])|@unless(?![a-zA-Z])|@switch(?![a-zA-Z])|@error(?![a-zA-Z])|@isset(?![a-zA-Z])|@empty(?![a-zA-Z])[^\S\r
]*\(|@guest(?![a-zA-Z])|@while(?![a-zA-Z])|@auth(?![a-zA-Z])|@for(?![a-zA-Z])|@if(?![a-zA-Z]))(?!                     # Not followed by:
                [^<]*               # ...
                (?<![?=-])          # ... (Make sure we don't confuse ?>, ->, and =>, with HTML opening tag closings)
                >                   # A ">" character that isn't preceded by a "<" character (meaning it's outside of a tag)
            )
```

removing leading indent spaces returns 954 results:
```
(@foreach(?![a-zA-Z])|@forelse(?![a-zA-Z])|@unless(?![a-zA-Z])|@switch(?![a-zA-Z])|@error(?![a-zA-Z])|@isset(?![a-zA-Z])|@empty(?![a-zA-Z])[^\S\r
]*\(|@guest(?![a-zA-Z])|@while(?![a-zA-Z])|@auth(?![a-zA-Z])|@for(?![a-zA-Z])|@if(?![a-zA-Z]))(?!                     # Not followed by:
                [^<]*               # ...
                (?<![?=-])          # ... (Make sure we don't confuse ?>, ->, and =>, with HTML opening tag closings)
                >                   # A ">" character that isn't preceded by a "<" character (meaning it's outside of a tag)
            )
```

removing leading indent spaces, comments, and new lines returns 753 results:
```
(@foreach(?![a-zA-Z])|@forelse(?![a-zA-Z])|@unless(?![a-zA-Z])|@switch(?![a-zA-Z])|@error(?![a-zA-Z])|@isset(?![a-zA-Z])|@empty(?![a-zA-Z])[^\S\r
]*\(|@guest(?![a-zA-Z])|@while(?![a-zA-Z])|@auth(?![a-zA-Z])|@for(?![a-zA-Z])|@if(?![a-zA-Z]))(?![^<]*(?<![?=-])>)
```

... so VS Code regex search is somewhat sensitive here ... but this gives me something to go on. This regex does find the "bug" shown above, but it does not find the following "intra-html-element" `@if` which *is* correctly handled (i.e., no morph marker inserted):

```
<li id="disclaimer-{{ $item->id }}" class="@if (!$loop->first) border-t border-gray-200 @endif">
```

So bottom line for me converting a mature site from v2 to v3 is I have close to 1K (that's just the openings!) morph marker insertion points that are maybe-kinda-sorta correctly evaluated and this all-critical regex pattern will likely continue to be tweaked in minor version upgrades .... hmmmmmm, I don't know about this.

**Comment by thumbtech** [comment]
Also FWIW regex101.com handles the PHP-generated pattern well!:

![Screenshot 2023-11-28 090516](https://github.com/livewire/livewire/assets/2287235/5962bd60-aa8c-408a-9219-72b6964a8175)

**Reply by Yinci** [reply]
I'd like to add this also happens within `script` tags, causing the scripts to fail since the HTML comment is not valid within JS (`Uncaught SyntaxError: expected expression, got '<'`).

**Comment by CodyPChristian** [comment]
Agreed on this issue as well, in the end I had to just disable the markers as they caused more harm than good. Most of the time if you have an issue requiring the use of the markers then its best to make sure you are not forgetting to set a wire:key somewhere, same goes if you get a comp not found JS error. Likely culprit is always missing wire:key lol. 😄
